js/main.js

```diff
@@ -4,21 +4,38 @@ const canvas = new Canvas("canvas", "canvas-container");
 window.addEventListener("load", () => {
     let selectButton = document.getElementById("select");
     let lineButton = document.getElementById("line");
+    let rectangleButton = document.getElementById("rectangle");
 
     selectButton.addEventListener("click", () => {
-        Path.removeEventListeners();
+        removeAllToolEventListeners();
         for (let i = 0; i < canvas.shapes.length; i++) {
             canvas.shapes[i].domElement.addEventListener("click", Shape.onSelect);
         }
     });
-    
+
     lineButton.addEventListener("click", () => {
-        canvas.clickToggle = true;
-        for (let i = 0; i < canvas.shapes.length; i++) {
-            let shape = canvas.shapes[i];
-            shape.domElement.removeEventListener("click", Shape.onSelect);
-            shape.hideNodes();
-        }
+        cleanUpOnSelectingNewTool();
         canvas.canvas.addEventListener("click", Path.onMouseClick);
     });
-});
\ No newline at end of file
+
+    rectangleButton.addEventListener("click", () => {
+        cleanUpOnSelectingNewTool();
+
+        canvas.canvas.addEventListener("click", Rectangle.onMouseClick);
+    })
+});
+
+function cleanUpOnSelectingNewTool() {
+    canvas.clickToggle = true;
```

Comment:
it seems fairly intensive to add and remove all these listeners all the time, if there are a lot of shapes?

I would try holding "activeShape(s)" in a canvas/MapGT package context (class variable) and have pass Map into each these shapes' constructors in order to provide each individual shape a way of registering a click handler in map scope. Then, on click on any shape, package level logic (such as what active shape/tool is) can decide the appropriate action to take - such as forwarding to the shape onClick, or activating new shapes, etc.

This is a bit vague, lmk if you want more detailing, or use your own judgment, if you think this is bad advice (have your own reasoning about what will take effort to maintain)

Reply:
I got what you're thinking; I will try to implement my understanding of it in the PR for the next tool if that's fine?